[docs] Add page for CSS variables support in @mui/material (#32050)
mui/material-ui: docs/data/material/experimental-api/css-variables/css-variables.md

> If you are curious to learn the benefits you would get by using CSS variables, check out the [RFC](https://github.com/mui/material-ui/issues/27651) for CSS variables support.

## Usage

At this moment, the API does not require any breaking changes for the Material UI components, however it depends on using a new experimental provider for the theme, called `Experimental_CssVarsProvider`.
Except for providing the theme in the inner React context, this new provider has as a responsibility to generate CSS variables out of all tokens in the theme that are not functions, and make them available in the context.
All these variables, are available under a key in the theme, named `vars`.
The structure of this object is identical to the theme structure, the only difference is that the values represent some CSS varaibles.

The best way to see this is by example.
Let's say we have this element that uses some theme token:

```jsx
const Example() {
  const theme = useTheme();
  return <div style={{ color: theme.palette.success.main }}>Success div</div>
}
```

{{"demo": "SuccessDivThemeToken.js", "hideToolbar": true }}

> If you try to inspect this element, you will see that the color style have directly a hex value, that comes some theme token:

Reviewer: ```suggestion
If you try to inspect it, you will see that the color style has a direct hex value, that comes from some theme token:
```

```html
<div style="color: #2e7d32;">...</div>
```

> On the other hand, if the div uses some CSS variable as in this example:The one using the CSS variables however, allows developers to track the path of where the token is defined in theme:

Reviewer: ```suggestion
Now, if the same div used CSS variables, as in the example below, you'll be able to track where that color token as been defined in the theme.
```

```jsx
const Example() {
  const theme = useTheme();
  return <div style={{ color: theme.vars.palette.success.main }}>Success div</div>
}
```

{{"demo": "SuccessDivCSSVariable.js", "hideToolbar": true }}

```html
<div style="color: var(--md-palette-success-main);">...</div>
```

> More over, if you inspect the styles, you can click on the CSS variable and see the actual color value.
> We believe this will significantly increase the DX when debugging the styles for the component.

Reviewer: ```suggestion
Additionally, by clicking on the CSS variable in the browser's dev tools, you'll be taken to the theme object where you'll see its direct hex value again. 
We believe this to be a significant improvement in developer experience when debugging styles for any component.
```

### Custom theme

> The example above uses the default theme. In the next example you will learn how to create your custom theme, using the `experimental_extendTheme` utility.
> 
> Before starting with the example, it's important to distingwish some differences in the theme structure.
> The new theme structure, lets you define all different themes you have in one theme, by defining different color schemes.
> Then, when you want to change the app to use different theme, you can use the `useColorScheme` to set the color scheme you want to use.
> This will cause the browser to re-write the CSS varibles to point to the new values, while the components won't change at all, as they should still link to the same variables).

Reviewer: ```suggestion
The example above is using the default Material UI theme. But, what if you wanted to create your own custom theme? Let's see how to do that using the `experimental_extendTheme` utility. 

To start, it is important to understand this new theme structure. A theme is a collection of color schemes, all defined within a single theme object. You'll be able to create more than light and dark color schemes if you want but for the sake of simplicity, let's use that as an example. Here's how you'd customize it:

```jsx
const theme = experimental_extendTheme({
  colorSchemes: {
    light: {
      palette: {
        primary: teal,
        secondary: deepOrange,
      },
    },
    dark: {
      palette: {
        primary: cyan,
        secondary: orange,
      },
    },
  },
});
```

If you want to be able to toggle between color schemes, you can use the `useColorScheme` hook. This will make the browser re-write the CSS variables, pointing them to the new values you've created. The components won't change at all, though—they should still refer to the same variables as before.
```

Enriching this section a bit more, it's very important! Though I'm still a bit confused with what the last sentence means: _"The components won't change at all, though—they should still refer to the same variables as before."._

Author: > Though I'm still a bit confused with what the last sentence means: _"The components won't change at all, though—they should still refer to the same variables as before."._

The component's styles are not changing at all (the same classname is applied, as none of the styles changed), only the values for the CSS variable changed.

What I mean with this is, previously it worked like this:

1. light mode:
.some-class {
  color: #123456
}
2. dark mode:
.some-other-class {
  color: #654321
}

Not it works like this:

1. light mode:
.some-class {
  color: var(--md-palette-primary-main); // #123456
}
2. dark mode:
.some-class {
  color: var(--md-palette-primary-main); // #654321
}

Reviewer: Alright, just pushed a refinement on this part! Thanks for explaining :)

{{"demo": "CssVarsCustomTheme.js", "iframe": true }}

> ### Customization

Reviewer: ```suggestion
### Customizing components
```

> At this moment, the `Button` component is the only component supporting CSS variables (the support is not requiring any breaking changes).
> We would appreciate any feedback you can give us, by trying to use this new API.
> If order to do that, you should wrap your application with a `Experimental_CssVarisProvider`, and try to customize the `Button` component using some CSS variables.

Reviewer: ```suggestion
As of now, the `Button` component is the only one supporting CSS variables (note that it doesn't cause any breaking change). 
To customize it using CSS variables, you'll need to wrap your application with `Experimental_CssVarisProvider`. 
Play around with the demo below! We'd appreciate any feedback about this API, as it is still in development. 
```

# CSS variables

<p class="description">Learn about the experimental API for using CSS variables on Material UI components.</p>

CSS variables is the most upvoted `@mui/system` related issue.
It allows several improvements in developer experience, like fixing dark mode SSR flickering, allowing multiple themes other than only light and dark, better debugging experience overall, and many others.

> To learn more, check [the RFC](https://github.com/mui/material-ui/issues/27651) we did about it.

Author: This was similar to what I initially said, but @siriwatknp said linking the RFC would be better. I am applying your changes, and we can revisit this before merge based on Jun's review :)

A while ago, we added an experimental API to the System package meant to allow styling the components with CSS variables.
We want now to take it one step further and utilize it in one of the Material UI components.

## Usage

> At this moment, the API does not require any breaking changes for the Material UI components, however it depends on using a new experimental provider for the theme, called `Experimental_CssVarsProvider`.

Author: Based on our experiments it shouldn't require braking changes, I hope it will stay like that in the future (at least we won't implement the bits that would potentially require braking changes till we have v6 alpha), but I really think there shouldn't be breaking changes.

Except for providing the theme in the inner React context, this new provider has as a responsibility to generate CSS variables out of all tokens in the theme that are not functions, and make them available in the context.

All these variables, are available under a key in the theme, named `vars`.
The structure of this object is identical to the theme structure, the only difference is that the values represent some CSS varaibles.

> The best way to see this is by example.

Reviewer 2: Cut this.

```suggestion

```

Let's say we have this element that uses some theme token:

```jsx
const Example() {
  const theme = useTheme();
  return <div style={{ color: theme.palette.success.main }}>Success div</div>
}
```

{{"demo": "SuccessDivThemeToken.js", "hideToolbar": true }}

> If you try to inspect it, you will see that the color style has a direct hex value, that comes from some theme token:

Reviewer 2: ```suggestion
If you inspect it, you will see that the color style has a direct hex value that comes from a theme token:
```

```html
<div style="color: #2e7d32;">...</div>
```

> 

Reviewer 2: I would elaborate on why this is a problem. This is based on my understanding after reading the RFC.

```suggestion

This makes the code difficult to debug, because the app could have multiple theme tokens, but this doesn't give you any information about which theme is responsible for this style.

```

> Now, if the same div used CSS variables, as in the example below, you'll be able to track where that color token is defined in the theme:

Reviewer 2: ```suggestion
If that same `<div>` used CSS variables instead—as in the example below—you could track _where_ that color token is defined in the theme:
```

```jsx
const Example() {
  const theme = useTheme();
  return <div style={{ color: theme.vars.palette.success.main }}>Success div</div>
}
```

{{"demo": "SuccessDivCSSVariable.js", "hideToolbar": true }}

```html
<div style="color: var(--md-palette-success-main);">...</div>
```

> Additionally, by clicking on the CSS variable in the browser's dev tools, you'll be taken to the theme object where you'll see its direct hex value again.

Reviewer 2: ```suggestion
Clicking on the CSS variable in the browser's dev tools takes you to the theme object where you'll see its direct hex value again.
```

> We believe this to be a significant improvement in developer experience when debugging styles for any component.

Reviewer 2: No need for weasel words—we *know* this is a big improvement! 😉

```suggestion
This offers a huge improvement to the overall experience of debugging component styles.
```

### Creating a custom theme

> The example above is using the default Material UI theme.
> But, what if you wanted to create your own custom theme?
> Let's see how to do that using the `experimental_extendTheme` utility.

Reviewer 2: Let's not assume that the reader has read the previous section. They might just be skimming.

```suggestion
If you want to override Material UI's default theme, you can create your own custom theme with CSS variables using the `experimental_extendTheme` utility.
```

To start, it is important to understand this new theme structure.

> A theme is a collection of color schemes, each one defined within a single theme object.

Reviewer 2: Rephrasing for clarity.

```suggestion
A theme is an object that contains a collection of color schemes.
```

> You'll be able to create more than light and dark color schemes if you want but, for the sake of simplicity, let's use only these two as an example. Here's how you'd customize it:

Reviewer 2: Rephrasing to separate distinct ideas.

```suggestion
Your custom theme can contain as many color schemes as you like, including light, dark and more.

Here's how to create light and dark schemes in a custom theme using CSS variables:
```

```jsx
const theme = experimental_extendTheme({
  colorSchemes: {
    light: {
      palette: {
        primary: teal,
        secondary: deepOrange,
      },
    },
    dark: {
      palette: {
        primary: cyan,
        secondary: orange,
      },
    },
  },
});
```

{{"demo": "CssVarsCustomTheme.js", "iframe": true }}

> To toggle between color schemes, you can use the `useColorScheme` hook.

Reviewer 2: Do we have a link or more info for this?

Author: Good point, we don't have at this moment. How about I add the API of the hook at the end of the page too, right after the `CssVarsProvider`'s API?

Author: Added

> This will make the browser re-write the CSS variables, pointing them to the new values you've created.
> The components' classes won't change at all, though—they should still refer to the same variables as before.

Reviewer 2: "Should" here seems to imply that they _might_ not actually behave as expected.

```suggestion
This will make the browser re-write the CSS variables, pointing them to the new values you created—but the components' classes will still refer to the same variables as before.
```

### Customizing components

> As of now, the `Button` component is the only one supporting CSS variables (note that it doesn't cause any breaking change).

Reviewer 2: Rephrasing.

Also, should this fact be closer to the top of the doc? I think I would be frustrated if I read through this whole doc and got excited to use this feature, but then discovered it only works with buttons currently.

```suggestion
Because the CSS variables API is an experimental feature, it is currently only supported by the `Button` component.
```

Author: Actually I agree with your comment here https://github.com/mui/material-ui/pull/32050/files#r839923202 I will add the note at the top of the page 👍

> To customize it using CSS variables, you'll need to wrap your application with `Experimental_CssVarisProvider`.

Reviewer 2: ```suggestion
To customize it using CSS variables, you'll need to wrap your application with `Experimental_CssVarsProvider`.
```

Play around with the demo below!
We'd appreciate any feedback about this API, as it is still in development.

{{"demo": "CssVariablesCustomization.js", "iframe": true }}

If you are using `TypeScript` you should use module augmentaiton to update the `Theme` structure:

```tsx
import { Theme as MuiTheme } from '@mui/material/styles';

declare module '@mui/material/styles' {
  interface Theme {
    vars: Omit<
      MuiTheme,
      'typography' | 'mixins' | 'breakpoints' | 'direction' | 'transitions'
    >;
  }
}
```

> > To help us bring this new API to more components faster, feel free to contribute by adding CSS variables support in more components. Make sure to check the [GitHub issue](https://github.com/mui/material-ui/issues/32049) that keeps track of our progress first, in case someone took the lead already on the component you were thinking of.

Reviewer 2: Rephrasing. Also, should this be near the top of the doc? I don't have a strong opinion but I could see it going either way.

```suggestion
> If you want to see wider support for this API across Material UI's component library, please feel free to contribute to the ongoing development. Make sure to check the [GitHub issue](https://github.com/mui/material-ui/issues/32049) that keeps track of our progress, to see if anyone else is currently working on a component you're interested in.
```

## API

`<CssVarsProvider>` props:

- `defaultColorScheme: ColorScheme | { light: ColorScheme; dark: ColorScheme }` - Design system default color scheme. Provides string if the design system has one default color scheme (either light or dark) or object if the design system has default light & dark color schemes.
- `defaultMode?: Mode` - Design system default mode.
- `disableTransitionOnChange : boolean` - Disable CSS transitions when switching between modes or color schemes
- `enableColorScheme: boolean` - Indicate to the browser which color scheme is used (light or dark) for rendering built-in UI
- `prefix: string` - CSS variable prefix

> - `theme: ThemeInput`

Author: Not sure how to describe it. It is the theme :D Maybe, "the theme provided to the React's context"?

# CSS variables

> <p class="description">Learn about the experimental API for using CSS variables on Material UI components.</p>

Reviewer 2: ```suggestion
<p class="description">Learn about the experimental API for using CSS variables with Material UI components.</p>
```

CSS variables provide significant improvements in developer experience related to theming and customization.

> With these variables, you can inject a theme into your app's stylesheet _at build time_ to apply the user's selected settings before the whole app is rendered.

Reviewer 2: Line break after.

```suggestion
With these variables, you can inject a theme into your app's stylesheet _at build time_ to apply the user's selected settings before the whole app is rendered.

```

Previously, these CSS variables were only available as an experimental API in the MUI System package.
Now they are ready for experimental use with Material UI components.

> If you want to see wider support for this API across Material UI's component library, please feel free to contribute to the ongoing development. Make sure to check the [GitHub issue](https://github.com/mui/material-ui/issues/32049) that keeps track of our progress, to see if anyone else is currently working on a component you're interested in.

## Introduction

The CSS variables API relies on a new experimental provider for the theme called `Experimental_CssVarsProvider` to inject styles into Material UI components.
In addition to providing the theme in the inner React context, this new provider also generates CSS variables out of all tokens in the theme that are not functions, and makes them available in the context as well.

All of these variables are accessible in an object in the theme called `vars`.
The structure of this object is nearly identical to the theme structure, the only difference is that the values represent CSS variables.

## Usage

> `Experimental_CssVarsProvider` is a new experimental provider that attach all generated CSS variables to the theme and put it in react context. The children under this provider will be able to read the CSS variables from the theme.

Reviewer: ```suggestion
`Experimental_CssVarsProvider` is a new experimental provider that attaches all generated CSS variables to the theme and put it in React context. Children elements under this provider will also be able to read the CSS variables from the theme.
```

```js
import { Experimental_CssVarsProvider as CssVarsProvider } from '@mui/material/styles';

function App() {
  return <CssVarsProvider>...</CssVarsProvider>;
}
```

### Toggle between light and dark mode

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Reviewer: ```suggestion
`Experimental_CssVarsProvider` provides light and dark mode by default. It stores the selected user mode and syncs it with the browser's local storage internally. You can read or update the mode using the `useColorScheme` API.

  Developers can read/update the mode via the `useColorScheme` API.
```

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Reviewer 2: I'm a little confused by this after re-reading the list of CssVarsProvider props. How does a "mode" differ from a "color scheme"? As I understand the terminology, light and dark "mode" are both types of "color schemes" (so these terms should be interchangeable) but I may not be familiar with the specific design jargon.

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Reviewer: This is something we had to figure out too for Joy. I'd say that my latest understanding of it is: mode is strictly either light or dark. But, each mode could have several color schemes. You can have several light mode palettes/color schemes and the same goes for dark. Does it make sense?

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Reviewer 2: That makes sense, but it seems kind of odd to me. So a dev can't create a "mode" (which could be one or more color schemes) unless it's specifically "light" or "dark"? Is a "mode" more like a "theme" then? Or it's like an organizational element that sits between the theme and the color scheme?

I think it's probably worthwhile to call out the terminology we're using here, because I'm not sure if the average docs reader would understand this distinction. Something like:

> Note: While the terms "color scheme" and "mode" may sometimes be used interchangeably, we use "mode" only when referring specifically to light and dark mode.
You may have one or more color schemes for each mode; you may also have color schemes that do not belong to any mode.

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Reviewer: Yeah, it's tricky 😅 And I'm pretty sure we don't have all that consistency about this topic in the docs. You'll see in our docs and elsewhere that theme and mode are also used interchangeably, both referring to dark and light modes. 

I see the theme as something that houses all the design direction—you'll find more stuff than palettes there, like spacing scale, font family, type scale, etc. Mode would be either dark or light (or auto, but that's not a third mode per se but rather the method as to how one or the other is chosen). Then, color schemes could be a synonym of palettes?! It refers to the number of schemes/palettes you have under one mode. 

Imagine you go to your computer and you choose dark mode (or appearance as Mac says). You could have several color schemes/palettes in a dark mode (fake examples: true black, grayish, cement, etc).

<img width="300" alt="Screen Shot 2022-04-04 at 15 11 43" src="https://user-images.githubusercontent.com/67129314/161606146-dd5b638a-5d4b-42c7-bad7-b617d202c881.png">

In summary, I'd say that only the last sentence is not accurate: _"you may also have color schemes that do not belong to any mode."_ Don't know how that could be. @siriwatknp your perspective here is very welcomed as we were discussing this in Joy's context as well!

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Reviewer 2: OK, so if—for example—I built an app that had four color palette options to choose from:

- light
- dark
- pink
- purple

I would have to classify "pink" and "purple" as color schemes that live within either light or dark mode?

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Reviewer: Yeah, you could have pink and purple on both light and dark modes, with the adequate adaptations: 

```
theme: {
   palettes: {
      light: {
        purple {...}
        pink: {...}
      }
       dark: {
        purple: {...}
        pink: {...}
     }
   }
}
```        
Does this clarifies?

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Reviewer 2: Yes, thank you! We should definitely explain this terminology in the doc here.

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Author: Sorry to chime in later, as Danilo mentioned mode so referring to `light`/`dark` or system (for example if the users has user preference for using dark mode in the browser.

Color schemes are different palettes that can be used.

The combination of these two produces the palette that will be used in the theme :)

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Author: Maybe an example can help. For example if there is a kids web app, you may have two different ambients for it: pink for girls and blue for boys. Both of them would support light and dark versions of them. Here is an example of how it can be implemented: https://codesandbox.io/s/bold-microservice-vgggz0?file=/src/App.js

> `Experimental_CssVarsProvider` provides light and dark mode by default. The provider stores the user selected mode and sync with the browser local storage internally. Developers can read/update the mode via the `useColorScheme` API.

Reviewer: Cool! We could use that example. But yeah, agree with Jun that all this explanation should probably be stored elsewhere and only referenced here in this doc.

In addition to providing the theme in the inner React context, this new provider also generates CSS variables out of all tokens in the theme that are not functions, and makes them available in the context as well.

All of these variables are accessible in an object in the theme called `vars`.
The structure of this object is nearly identical to the theme structure, the only difference is that the values represent CSS variables.

## Usage

`Experimental_CssVarsProvider` is a new experimental provider that attaches all generated CSS variables to the theme and puts them in React's context. Children elements under this provider will also be able to read the CSS variables from the theme.

```js
import { Experimental_CssVarsProvider as CssVarsProvider } from '@mui/material/styles';

function App() {
  return <CssVarsProvider>...</CssVarsProvider>;
}
```

### Customizing components

Because the CSS variables API is an experimental feature, it is currently only supported by the `Button` component.
To customize it using CSS variables, you'll need to wrap your application with `Experimental_CssVarsProvider`.

Play around with the demo below!

> We'd appreciate any feedback about this API, as it is still in development.

Reviewer: I think I'd remove this bit here, it's weird to be only inside the "Customizing components" sub-section.
What I'd maybe do is add a "Feedback" topic, below the "API" one, and add more instructions as to how to give feedback?! Where to open the issue, how to do it, how to structure the writing, etc.

Author: Yeah makes sense. It's probably better to add it before the API actually, as I think the API is the last section always on a pages where it is next to the content.

Author: I actually move this sentence in the top Note, feels like it belongs there, as we are already asking for help, asking for feedback can be complementary

{{"demo": "CssVariablesCustomization.js", "iframe": true }}

If you are using TypeScript you should use module augmentation to update the `Theme` structure:

```tsx
import { Theme as MuiTheme } from '@mui/material/styles';

declare module '@mui/material/styles' {
  interface Theme {
    vars: Omit<
      MuiTheme,
      'typography' | 'mixins' | 'breakpoints' | 'direction' | 'transitions'
    >;
  }
}
```

### Customizing the theme

> If you want to override Material UI's default color schemes, you can use the `experimental_extendTheme` utility.

Reviewer: ```suggestion
If you want, for example, to override Material UI's default color schemes, you can use the `experimental_extendTheme` utility.
```

```jsx
const theme = experimental_extendTheme({
  colorSchemes: {
    light: {
      palette: {
        primary: teal,
        secondary: deepOrange,
      },
    },
    dark: {
      palette: {
        primary: cyan,
        secondary: orange,
      },
    },
  },
});
```

{{"demo": "CssVarsCustomTheme.js", "iframe": true }}

If you are using [`ThemeProvider`](/material-ui/customization/theming/#theme-provider), you can replace it with the new experimental provider.

```diff
- import { ThemeProvider, createTheme } from '@mui/material/styles';
+ import { Experimental_CssVarsProvider as CssVarsProvider } from '@mui/material/styles';

function App() {
  return (
-    <ThemeProvider theme={createTheme()}>
-      ...
-    </ThemeProvider>
+    <CssVarsProvider>
+      ...
+    </CssVarsProvider>
  )
}
```

### Toggle between light and dark mode

`Experimental_CssVarsProvider` provides light and dark mode by default.
It stores the selected user mode and syncs it with the browser's local storage internally.

> Developers can read/update the mode via the `useColorScheme` API.

Reviewer: ```suggestion
You can read and update the mode via the `useColorScheme` API.
```

```jsx
import {
  Experimental_CssVarsProvider as CssVarsProvider,
  useColorScheme,
} from '@mui/material/styles';

const ModeSwitcher = () => {
  const { mode, setMode } = useColorScheme();
  const [mounted, setMounted] = React.useState(false);

  React.useEffect(() => {
    setMounted(true);
  }, []);

  if (!mounted) {
    // for server-side rendering
    // Read more on https://github.com/pacocoursey/next-themes#avoid-hydration-mismatch
    return null;
  }

  return (
    <Button
      variant="outlined"
      onClick={() => {
        if (mode === 'light') {
          setMode('dark');
        } else {
          setMode('light');
        }
      }}
    >
      {mode === 'light' ? 'Dark' : 'Light'}
    </Button>
  );
};

function App() {
  return (
    <CssVarsProvider>
      <ModeSwitcher />
    </CssVarsProvider>
  );
}
```

### Server-side rendering

> To prevent the dark-mode SSR flickering during the hydration phase, place `getInitColorSchemeScript()` before the `<Main />` tag.

Reviewer: A code block here would be helpful!

Author: There are both for nextjs and gatsby in the following sections.

Reviewer: Oh, okay. So I guess that the Next.js and Gatsby should be #### (nested within the Server-side rendering topic), no?!

Author: Yes, exactly!

> ### Next JS

Reviewer: ```suggestion
### Next.js
```

> ### Next.js

Reviewer: ```suggestion
#### Next.js
```

> Add the following code to `pages/_document.js`.

Reviewer: ```suggestion
To use the API with a Next.js project, add the following code to the `pages/_document.js` file.
```

```jsx
import Document, { Html, Head, Main, NextScript } from 'next/document';
import { getInitColorSchemeScript } from '@mui/material/styles';

export default class MyDocument extends Document {
  render() {
    return (
      <Html>
        <Head>...</Head>
        <body>
          {getInitColorSchemeScript()}
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}
```

### Gatsby

> Add the following code to `gatsby-ssr.js`

Reviewer: ```suggestion
To use the API with a Gatsby project, add the following code to the `gatsby-ssr.js file`
```

> ### Gatsby

Reviewer: ```suggestion
#### Gatsby
```

It stores the selected user mode and syncs it with the browser's local storage internally.
You can read and update the mode via the `useColorScheme` API.

```jsx
import {
  Experimental_CssVarsProvider as CssVarsProvider,
  useColorScheme,
} from '@mui/material/styles';

const ModeSwitcher = () => {
  const { mode, setMode } = useColorScheme();
  const [mounted, setMounted] = React.useState(false);

  React.useEffect(() => {
    setMounted(true);
  }, []);

  if (!mounted) {
    // for server-side rendering
    // Read more on https://github.com/pacocoursey/next-themes#avoid-hydration-mismatch
    return null;
  }

  return (
    <Button
      variant="outlined"
      onClick={() => {
        if (mode === 'light') {
          setMode('dark');
        } else {
          setMode('light');
        }
      }}
    >
      {mode === 'light' ? 'Dark' : 'Light'}
    </Button>
  );
};

function App() {
  return (
    <CssVarsProvider>
      <ModeSwitcher />
    </CssVarsProvider>
  );
}
```

### Server-side rendering

To prevent the dark-mode SSR flickering during the hydration phase, place `getInitColorSchemeScript()` before the `<Main />` tag.

#### Next.js

> To use the API with a Next.js project, add the following code to the `pages/_document.js` file.

Reviewer 2: Colon. Also, I think it's worth adding a reference for the custom `_document` since this is an advanced feature (and some users might be confused when they can't find it in their fresh app).

```suggestion
To use the API with a Next.js project, add the following code to the custom [`pages/_document.js`](https://nextjs.org/docs/advanced-features/custom-document) file:
```

```jsx
import Document, { Html, Head, Main, NextScript } from 'next/document';
import { getInitColorSchemeScript } from '@mui/material/styles';

export default class MyDocument extends Document {
  render() {
    return (
      <Html>
        <Head>...</Head>
        <body>
          {getInitColorSchemeScript()}
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}
```

#### Gatsby

> To use the API with a Gatsby project, add the following code to the `gatsby-ssr.js file`

Reviewer 2: Same idea as the Next.js suggestion, since this file is something that's not included in a blank Gatsby app.

```suggestion
To use the API with a Gatsby project, add the following code to the custom [`gatsby-ssr.js](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-ssr/) file`:
```

```jsx
import React from 'react';
import { getInitColorSchemeScript } from '@mui/material/styles';

export function onRenderBody({ setPreBodyComponents }) {
  setPreBodyComponents([getInitColorSchemeScript()]);
}
```

## API

### `<CssVarsProvider>` props

- `defaultMode?: 'light' | 'dark' | 'system'` - Application's default mode (`light` by default)
- `disableTransitionOnChange : boolean` - Disable CSS transitions when switching between modes
- `enableColorScheme: boolean` - Indicate to the browser which color scheme is used (light or dark) for rendering built-in UI
- `prefix: string` - CSS variable prefix

> - `theme: ThemeInput` - the theme provided to the React's context

Reviewer 2: ```suggestion
- `theme: ThemeInput` - the theme provided to React's context
```

Play around with the demo below!
We'd appreciate any feedback about this API, as it is still in development.

{{"demo": "CssVariablesCustomization.js", "iframe": true }}

If you are using TypeScript you should use module augmentation to update the `Theme` structure:

```tsx
import { Theme as MuiTheme } from '@mui/material/styles';

declare module '@mui/material/styles' {
  interface Theme {
    vars: Omit<
      MuiTheme,
      'typography' | 'mixins' | 'breakpoints' | 'direction' | 'transitions'
    >;
  }
}
```

## API

### `<CssVarsProvider>` props

- `defaultColorScheme: ColorScheme | { light: ColorScheme; dark: ColorScheme }` - Design system default color scheme; provides a string if the design system has one default color scheme (either light or dark) or an object if the design system has default light & dark color schemes
- `defaultMode?: Mode` - Design system default mode
- `disableTransitionOnChange : boolean` - Disable CSS transitions when switching between modes or color schemes
- `enableColorScheme: boolean` - Indicate to the browser which color scheme is used (light or dark) for rendering built-in UI
- `prefix: string` - CSS variable prefix

> - `theme: ThemeInput`

Reviewer 3: ```suggestion
- `theme: ThemeInput` - the theme provided to the React's context
```
